sparse/coo/core.py

```diff
@@ -372,14 +350,86 @@ def from_scipy_sparse(cls, x):
         >>> np.array_equal(x.todense(), s.todense())
         True
         """
-        x = scipy.sparse.coo_matrix(x)
+        x = x.asformat('coo')
         coords = np.empty((2, x.nnz), dtype=x.row.dtype)
         coords[0, :] = x.row
         coords[1, :] = x.col
         return COO(coords, x.data, shape=x.shape,
                    has_duplicates=not x.has_canonical_format,
                    sorted=x.has_canonical_format)
 
+    @classmethod
+    def from_iter(cls, x, shape=None):
+        """
+        Converts an iterable in certain formats to a :obj:`COO` array. See examples
+        for details.
+
+        Parameters
+        ----------
+        x : dict or list or tuple
+            The iterable to convert to :obj:`COO`.
+        shape : tuple[int], optional
+            The shape of the array.
+
+        Returns
+        -------
+        out : COO
+            The output :obj:`COO` array.
+
+        Examples
+        --------
+        You can convert items of the format ``[((i, j, k), value), ((i, j, k), value)]`` to :obj:`COO`.
+        Here, the first part represents the coordinate and the second part represents the value.
+
+        >>> x = [((0, 0), 1), ((1, 1), 1)]
+        >>> s = COO.from_iter(x)
+        >>> s.todense()
+        array([[1, 0],
+               [0, 1]])
+
+        You can also have a similar format with a dictionary.
+
+        >>> x = {(0, 0): 1, (1, 1): 1}
+        >>> s = COO.from_iter(x)
+        >>> s.todense()
+        array([[1, 0],
+               [0, 1]])
+
+        The third supported format is ``(data, (..., row, col))``.
+
+        >>> x = ([1, 1], ([0, 1], [0, 1]))
+        >>> s = COO.from_iter(x)
+        >>> s.todense()
+        array([[1, 0],
+               [0, 1]])
+        """
+        if isinstance(x, dict):
+            x = list(x.items())
+
+        if len(x) != 2 and not all(len(item) == 2 for item in x):
+            raise ValueError('Invalid iterable to convert to COO.')
```

Comment:
Is it possible that they user will give an `Iterator` rather than an `Iterable`?  If so then this check will consume our data.  We might want to defensively convert to a tuple or list.

Reply:
I'm not too familiar with the differences between the two. Will this check break or is it just a performance consideration?

Reply:
Iterators are ephemeral and get consumed on first use.  They are common, for example, when reading a sequence of data from a file.  Here is an example that creates an iterator (seq) from an iterable (L) does a check, and then discovers that some values have been removed:

```
In [1]: L = [1, 2, 3]

In [2]: seq = iter(L)

In [3]: if all(x > 100 for x in seq):
   ...:     print("hello")
   ...:     

In [4]: list(seq)
Out[4]: [2, 3]

In [5]: from collections import Iterator, Iterable

In [7]: isinstance(seq, Iterator)
Out[7]: True

In [8]: isinstance(L, Iterable)
Out[8]: True

In [9]: L
Out[9]: [1, 2, 3]
```

Reply:
Ah, I understand now. I don't think making this a supported input is worth the extra effort.

Reply:
I suspect that users will use iterators here whether or not we support it.  Currently we silently fail.  We should either fail loudly

```python
if isinstance(x, Iterator):
    raise TypeError("...")
```

Or we should coerce to a list

```python
if isinstance(x, Iterator):
    x = list(x)
```

Reply:
Well, if we have to add code, I guess supporting it is better, so I did that.